File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        ans = 0
        l = 0
        r = n-1
        left_max = -1
        right_max = -1

        for i in range(1,n):
            if height[l] <= height[r]:
                if height[l] > left_max:
                    left_max = height[l]
                else:
                    ans += left_max-height[l]
                l += 1
            else:
                if height[r] > right_max:
                    right_max = height[r]
                else:
                    ans += right_max-height[r]
                r -= 1
                
        return ans
                

        

File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int water = 0;
        int n = height.length;
        int left = 0;
        int right = n-1;
        int leftMax = height[0];
        int rightMax = height[n-1];

        while(left < right){
            if(leftMax <= rightMax){
                water += Math.min(leftMax, rightMax) - height[left];
                left++;
                leftMax = Math.max(leftMax, height[left]);
            }
            else{
                water += Math.min(leftMax, rightMax) - height[right];
                right--;
                rightMax = Math.max(rightMax, height[right]);
            }
        }

        return water;
    }
}